tools/sim/bridge/metadrive.py

```diff
@@ -0,0 +1,126 @@
+        o, _, d, _ = self.env.step(vc)
+        if d:
+          print("!!!Episode terminated due to violation of safety!!!")
+          self.env.reset()
+          self.env.step([0.0, 0.0])
+          for _ in range(300):
+            rk.keep_time()
+        
+        self.speed = o["state"][3] * math.sqrt(self.ticks_per_frame) * 10 # empirically derived
+
+      img = self.env.vehicle.image_sensors["rgb_wide"].get_pixels_array(self.env.vehicle, False)
+      self.yuv = self.camerad.img_to_yuv(img)
+      self.camerad.cam_send_yuv_wide_road(self.yuv)
+
+  def get_velocity(self):
+    return None
+
+  def get_speed(self) -> float:
+    return self.speed
+
+  def get_steer_correction(self) -> float:
+    max_steer_angle = 75 / self.ticks_per_frame
+    return max_steer_angle * STEER_RATIO * -1
+  
+  def tick(self):
+    pass
+
+
+class MetaDriveBridge(SimulatorBridge):
+  TICKS_PER_FRAME = 10
+
+  def __init__(self, args):
+    if args.dual_camera:
+      warnings.warn("Dual camera not supported in MetaDrive simulator for performance reasons")
+      args.dual_camera = False
+    if args.ticks_per_frame:
+      self.TICKS_PER_FRAME = args.ticks_per_frame
+    super(MetaDriveBridge, self).__init__(args)
+
+  def spawn_objects(self):
+    env = gym.make('MetaDrive-10env-v0', config=dict(offscreen_render=True))
+
+    # config = dict(
+    #   # camera_dist=3.0,
+    #   # camera_height=1.0,
+    #   # use_render=True,
+    #   vehicle_config=dict(
+    #     # enable_reverse=True,
+    #     # image_source="rgb_camera",
+    #     # rgb_camera=(0,0)
+    #   ),
+    #   offscreen_render=True,
+    # )  
+    # from metadrive.utils.space import VehicleParameterSpace
+    # from metadrive.component.vehicle.vehicle_type import DefaultVehicle
+    # max_engine_force = VehicleParameterSpace.DEFAULT_VEHICLE["max_engine_force"]
+    # max_engine_force._replace(max=max_engine_force.max * 10)
+    # max_engine_force._replace(min=max_engine_force.min * 10)
+
+    env.reset()
+    from metadrive.constants import CamMask
+    from metadrive.component.vehicle_module.base_camera import BaseCamera
+    from metadrive.engine.engine_utils import engine_initialized
+    from metadrive.engine.core.image_buffer import ImageBuffer
+    class RGBCameraWide(BaseCamera):
+      # shape(dim_1, dim_2)
+      BUFFER_W = W  # dim 1
+      BUFFER_H = H  # dim 2
+      CAM_MASK = CamMask.RgbCam
+
+      def __init__(self):
+        assert engine_initialized(), "You should initialize engine before adding camera to vehicle"
+        self.BUFFER_W, self.BUFFER_H = W, H
+        super(RGBCameraWide, self).__init__()
+        cam = self.get_cam()
+        lens = self.get_lens()
+        cam.lookAt(0, 2.4, 1.3)
+        cam.setHpr(0, 0.8, 0)
+        lens.setFov(160)
+        lens.setAspectRatio(1.15)
+  
+    env.vehicle.add_image_sensor("rgb_wide", RGBCameraWide())
+
+    # Monkey patch `get_rgb_array` since it involves unnecessary copies
+    def patch_get_rgb_array(self):
```

Comment:
After profiling, turns out the np transformations are _very, very slow_

Reply:
getScreenshot took 7.489ms
pixel array took 28.336ms

Reply:
further, `getScreenshot` is the majority component of `step`, so if we can disable on a normal step, perhaps we can step more frequently than generating the image.

Reply:
You don't need get pixel array with my method

Reply:
Under the hood, you are still using `get_rgb_array`. That's how the image obs is defined.

Reply:
I just do getscreenshot() and use it in another thread

Reply:
Oops, I made a mistake, Ive now fixed the actually expensive part of the function call.